packages/pigeon/lib/dart_generator.dart

```diff
@@ -11,66 +11,83 @@ class DartOptions {
   assert(api.location == ApiLocation.host);
-  final String nullTag = opt.isNullSafe ? '?' : '';
+  final String nullable = opt.isNullSafe ? '?' : '';
+  final String notNullable = opt.isNullSafe ? '!' : '';
+  bool first = true;
   indent.write('class ${api.name} ');
   indent.scoped('{', '}', () {
     for (Method func in api.methods) {
+      if (!first) {
+        indent.writeln('');
+      } else {
+        first = false;
+      }
       String argSignature = '';
       String sendArgument = 'null';
-      String requestMapDeclaration;
+      String encodedDeclaration;
       if (func.argType != 'void') {
         argSignature = '${func.argType} arg';
-        sendArgument = 'requestMap';
-        requestMapDeclaration =
-            'final Map<dynamic, dynamic> requestMap = arg._toMap();';
+        sendArgument = 'encoded';
+        encodedDeclaration = 'final Object encoded = arg.encode();';
       }
       indent.write(
-          'Future<${func.returnType}> ${func.name}($argSignature) async ');
+        'Future<${func.returnType}> ${func.name}($argSignature) async ',
+      );
       indent.scoped('{', '}', () {
-        if (requestMapDeclaration != null) {
-          indent.writeln(requestMapDeclaration);
+        if (encodedDeclaration != null) {
+          indent.writeln(encodedDeclaration);
         }
         final String channelName = makeChannelName(api, func);
-        indent.writeln('const BasicMessageChannel<dynamic> channel =');
-        indent.inc();
-        indent.inc();
-        indent.writeln(
-            'BasicMessageChannel<dynamic>(\'$channelName\', StandardMessageCodec());');
-        indent.dec();
-        indent.dec();
-        indent.writeln('');
+        indent.writeln('const BasicMessageChannel<Object$nullable> channel =');
+        indent.nest(2, () {
+          indent.writeln(
+            'BasicMessageChannel<Object$nullable>(\'$channelName\', StandardMessageCodec());',
+          );
+        });
         final String returnStatement = func.returnType == 'void'
             ? '// noop'
-            : 'return ${func.returnType}._fromMap(replyMap[\'${Keys.result}\']);';
+            : 'return ${func.returnType}.decode(replyMap[\'${Keys.result}\']$notNullable);';
         indent.format(
-            '''final Map<dynamic, dynamic>$nullTag replyMap = await channel.send($sendArgument);
+            '''final Map<Object$nullable, Object$nullable>$nullable replyMap = await channel.send($sendArgument) as Map<Object$nullable, Object$nullable>$nullable;
 if (replyMap == null) {
 \tthrow PlatformException(
 \t\tcode: 'channel-error',
 \t\tmessage: 'Unable to establish connection on channel.',
-\t\tdetails: null);
+\t\tdetails: null,
+\t);
 } else if (replyMap['error'] != null) {
-\tfinal Map<dynamic, dynamic> error = replyMap['${Keys.error}'];
+\tfinal Map<Object$nullable, Object$nullable> error = replyMap['${Keys.error}'] as Map<Object$nullable, Object$nullable>;
```

Comment:
`dynamic` hides bugs, because it disables all type checking. By using `Object?` we can see where we're casting and we can make sure the methods (etc) we're using on those objects don't have typos.

Reply:
(To answer your questions, `dynamic` is essentially the same as `Object?`, except that the language never checks what you call on objects that are marked `dynamic`, it's like you're in an untyped language. Everything is an `Object?`.)